test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/ApmIT.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.test.apmintegration;
+
+import org.elasticsearch.client.Request;
+import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.local.distribution.DistributionType;
+import org.elasticsearch.test.rest.ESRestTestCase;
+import org.hamcrest.Matchers;
+import org.junit.ClassRule;
+import org.junit.Rule;
+
+import static org.hamcrest.Matchers.containsString;
+
+public class ApmIT extends ESRestTestCase {
+    @ClassRule
+    public static RecordingApmServer mockApmServer = new RecordingApmServer();
+
+    @Rule
+    public ElasticsearchCluster cluster = ElasticsearchCluster.local()
+        .distribution(DistributionType.INTEG_TEST)
+        .module("test-apm-integration")
+        .module("apm")
+        .setting("telemetry.metrics.enabled", "true")
+        .setting("tracing.apm.agent.metrics_interval", "1s")
+        .setting("tracing.apm.agent.server_url", "http://127.0.0.1:" + mockApmServer.getPort())
+        .build();
+
+    @Override
+    protected String getTestRestCluster() {
+        return cluster.getHttpAddresses();
+    }
+
+    public void testApmIntegration() throws Exception {
+        client().performRequest(new Request("GET", "/_use_apm_metrics"));
+
+        // TestMeterUsages for code that is registering & using metrics
+        assertBusy(() -> {
+            var usageRecords = mockApmServer.getMessages();
```

Comment:
I think we would still need some kind of polling. the recording apm server is also receiving other 'default' metrics like those from GC.
when running locally those metrics are merged by apm agent and sent as a one json message to apm server.
However it could happen that in CI (or prod) those metrics  would be in separate messages sent to apm server

Reply:
I think we can do it without polling. A listener can be attached to read each new message to the server. Then have that listener decrement a latch when the expected message is received. In the main test thread, wait on the latch.

Reply:
My suggestion is specifically about filtering in an efficient way. Right now, all messages are checked in a bundle. If any of the assertions fail, another check is run after a short while, and all the same checks are done, even ones that have already passed. Instead, a listener could effectively check off each expected message, and do so without polling.

Reply:
I am not sure efficient filtering is a problem in this test. We are expecting 6 messages after all.
You probably have in mind something like  MockLogAppender.assertThatLogger method?

I am not sure this will make this code more readable and deterministic though. It would make the testing code more complex.
I am not against the idea, but in this test I think it might be an overdo. Perhaps if we were to apply this to other usages of 'simple http server with expectations' it would be worth a refactor? We could use this in Metering and Tracing too

Reply:
after speaking with @rjernst offline we concluded that assertBusy is potentially wasting a lot of time (it has an exponential backoff mechanism).
I created a draft of a refactoring to implement the idea with a listener, I wasn't sure on the results though.
Take a look here https://github.com/pgomulka/elasticsearch/pull/34